src/category_theory/connected.lean

```diff
@@ -0,0 +1,336 @@
+  connected J :=
+connected.of_constant_of_preserves_morphisms (λ α F a, h {j | F j = F (default J)} rfl (λ _ _ f, by simp [a f] ))
+
+/-- j₁ and j₂ are related by `zag` if there is a morphism between them. -/
+@[reducible]
+def zag (j₁ j₂ : J) : Prop := nonempty (j₁ ⟶ j₂) ∨ nonempty (j₂ ⟶ j₁)
+/--
+`j₁` and `j₂` are related by `zigzag` if there is a chain of
+morphisms from `j₁` to `j₂`, with backward morphisms allowed.
+-/
+@[reducible]
+def zigzag : J → J → Prop := relation.refl_trans_gen zag
+
+/-- Any equivalence relation containing (⟶) holds for all pairs of a connected category. -/
+lemma equiv_relation [connected J] (r : J → J → Prop) (hr : _root_.equivalence r)
+  (h : ∀ {j₁ j₂ : J} (f : j₁ ⟶ j₂), r j₁ j₂) :
+  ∀ (j₁ j₂ : J), r j₁ j₂ :=
+begin
+  have z: ∀ (j : J), r (default J) j :=
+    induct_on_objects (λ k, r (default J) k)
+        (hr.1 (default J)) (λ _ _ f, ⟨λ t, hr.2.2 t (h f), λ t, hr.2.2 t (hr.2.1 (h f))⟩),
+  intros, apply hr.2.2 (hr.2.1 (z _)) (z _)
+end
+
+/-- In a connected category, any two objects are related by `zigzag`. -/
+lemma connected_zigzag [connected J] (j₁ j₂ : J) : zigzag j₁ j₂ :=
+equiv_relation _
+  (mk_equivalence _
+    relation.reflexive_refl_trans_gen
+    (relation.refl_trans_gen.symmetric (λ _ _ _, by rwa [zag, or_comm]))
+    relation.transitive_refl_trans_gen)
+  (λ _ _ f, relation.refl_trans_gen.single (or.inl (nonempty.intro f))) _ _
+
+/--
+If any two objects in an inhabited category are related by `zigzag`, the category is connected.
+-/
+def zigzag_connected [inhabited J] (h : ∀ (j₁ j₂ : J), zigzag j₁ j₂) : connected J :=
+begin
+  apply connected.of_induct,
+  intros,
+  have: ∀ (j₁ j₂ : J), zigzag j₁ j₂ → (j₁ ∈ p ↔ j₂ ∈ p),
+    introv k,
+    apply relation.refl_trans_gen.head_induction_on k,
+    refl,
+    intros,
+    rw ← a_3,
+    rcases h' with ⟨⟨_⟩⟩ | ⟨⟨_⟩⟩,
+    apply a_1 h',
+    symmetry,
+    apply a_1 h',
+  rwa this j (default J) (h _ _),
+end
+
+/--
+If J is connected, for any two objects there is a sequence of morphisms (some reversed) from one
+to the other.
+
+Converse is given in `connected_of_zigzag`.
+-/
+lemma exists_zigzag' [connected J] (j₁ j₂ : J) :
+  ∃ (l : list J), list.chain' zag l ∧ ∃ (hl : l ≠ []), list.head_of_ne_nil l hl = j₁ ∧ list.last l hl = j₂ :=
+list.exists_zigzag (connected_zigzag _ _)
+
+/--
+If any two objects in an inhabited category are linked by a sequence of (potentially reversed)
+morphisms, then J is connected.
+
+The converse of `exists_zigzag'`.
+-/
+def connected_of_zigzag [inhabited J]
+  (h : ∀ (j₁ j₂ : J), ∃ (l : list J), list.chain' zag l ∧ ∃ (hl : l ≠ []), list.head_of_ne_nil l hl = j₁ ∧ list.last l hl = j₂) :
+  connected J :=
+begin
+  apply connected.of_induct,
+  intros p d k j,
+  obtain ⟨l, zags, nemp, hd, tl⟩ := h j (default J),
+  apply list.prop_up_chain' p l nemp zags hd tl _ d,
+  rintros _ _ (⟨⟨_⟩⟩ | ⟨⟨_⟩⟩),
+  { exact (k a).2 },
+  { exact (k a).1 }
+end
+
+end J
+
+section examples
```

Comment:
I propose moving the examples, and `prod_preserves_connected_limits`, into `category_theory/limits/connected.lean`, and significantly reducing the imports of this file.